hazelcast/src/main/java/com/hazelcast/sql/impl/exec/fetch/Fetch.java

```diff
@@ -0,0 +1,146 @@
+
+    private final Expression<?> fetch;
+    private final Expression<?> offset;
+
+    private long fetchValue;
+    private long offsetValue;
+
+    private long offsetApplied;
+    private long fetchApplied;
+
+    public Fetch(Expression<?> fetch, Expression<?> offset) {
+        this.fetch = fetch;
+        this.offset = offset;
+    }
+
+    public void setup(ExpressionEvalContext context) {
+        fetchValue = fetch == null ? Long.MAX_VALUE : eval(fetch, true, context);
+        offsetValue = offset == null ? 0 : eval(offset, false, context);
+    }
+
+    public RowBatch apply(RowBatch batch) {
+        if (fetchApplied == fetchValue) {
+            return EmptyRowBatch.INSTANCE;
+        }
+
+        long toOffset = offsetValue - offsetApplied;
+
+        if (toOffset >= batch.getRowCount()) {
+            // Skip the whole batch
+            offsetApplied += batch.getRowCount();
+
+            return EmptyRowBatch.INSTANCE;
+        }
+
+        // Mark offset as finished and adjust the start index accordingly.
+        int startIndex;
+
+        if (toOffset != 0L) {
+            offsetApplied = offsetValue;
+
+            // toOffset is smaller than the batch size and casting to int is safe
+            startIndex = (int) toOffset;
+        } else {
+            startIndex = 0;
+        }
+
+        // Calculate the end position.
+        long toFetch = fetchValue - fetchApplied;
+        long canFetch = batch.getRowCount() - startIndex;
+
+        if (toFetch >= canFetch) {
+            fetchApplied += canFetch;
+
+            // Return the whole remainder.
+            if (startIndex == 0) {
+                return batch;
+            } else {
+                return slice(batch, startIndex, (int) canFetch);
+            }
+        } else {
+            fetchApplied += toFetch;
+
+            return slice(batch, startIndex, (int) toFetch);
+        }
+    }
+
+    public boolean isDone() {
+        return fetchApplied == fetchValue;
+    }
+
+    private long eval(Expression<?> expression, boolean fetch, ExpressionEvalContext context) {
+        assert expression != null;
+
+        String name = fetch ? "LIMIT" : "OFFSET";
+
+        Object val = expression.eval(FetchRow.INSTANCE, context);
+
+        if (val == null) {
+            throw QueryException.error("Value of " + name + " cannot be null");
+        }
+
+        long val0;
+
+        if (val instanceof Number) {
+            val0 = ((Number) val).longValue();
+        } else {
+            try {
+                val0 = Converters.getConverter(val.getClass()).asBigint(val);
+            } catch (Exception e) {
+                throw QueryException.error("Cannot convert " + name + " value to number: " + val, e);
+            }
+        }
+
+        if (val0 < 0L) {
+            throw QueryException.error("Value of " + name + " cannot be negative: " + val);
+        }
+
+        return val0;
+    }
+
+    private RowBatch slice(RowBatch source, int start, int len) {
```

Comment:
The implementation of this method appears to be inefficient because it adds copying of the row-batch. It is likely there is a way to avoid copying completely.

Reply:
We can add the `slice` method to the RowBatch. It will be something like 

`    RowBatch slice(int start, int len);
`
@devozerov , Is that what you meant?

Reply:
I mean, that when we doing a sort-merge, there is no need to construct the batch without LIMIT/OFFSET, and then apply LIMIT/OFFSET separately, producing the new batch. Instead, the code may be reorganized in a way, that the sorter filters out rows on its own, thus avoiding an intermediate copy.

Reply:
Yes, this part is clear. However, the `Fetch` processor is also used without sort-merge. And my question was do we want to avoid copying on `slice` as well?

Reply:
Integrated FETCH/OFFSET into the merge sorter in https://github.com/hazelcast/hazelcast/pull/18091/commits/3cc58e602a8a164202f8e834241d2bc09f055e8f